Fix mismatch function argument between cassandra_cql and cassandra_cql_return

GwiYeong opened the pull request:
### What does this PR do?
there is mismatch keyword argument between `cassandra_cql` and `cassandra_cql_return`

the function [cql_query_with_prepare](https://github.com/saltstack/salt/blob/master/salt/modules/cassandra_cql.py#L365) is defined like below

```python
def cql_query_with_prepare(query, statement_name, statement_arguments, callback_errors=None, contact_points=None, port=None, cql_user=None, cql_pass=None, **kwargs):

    # Backward-compatibility with Python 3.7: "async" is a reserved word
    asynchronous = kwargs.get('async', False)
```

the first line of the function gets `async` param from `**kwargs`. 

however the returner, `cassandra_cql_return` calls this function with argument that is named `asynchronous`.

See the below code line.

https://github.com/saltstack/salt/blob/master/salt/returners/cassandra_cql_return.py#L205
https://github.com/saltstack/salt/blob/master/salt/returners/cassandra_cql_return.py#L226
https://github.com/saltstack/salt/blob/master/salt/returners/cassandra_cql_return.py#L268
https://github.com/saltstack/salt/blob/master/salt/returners/cassandra_cql_return.py#L298

these functions call `cassandra_cql.cql_query_with_prepare` with `asynchronous=True`, but actually `cql_query_with_prepare` not work asynchronous

I found this commit 377fbbc7b760bad1fd267747bf463e7cc8f3c1ae. I could not understand this commit. 
since python 3.7, `async` is reserved keyword, so use `asynchronous` for function args looks like correct.  but this commit remove `asynchronous` arg.
therefore I revert that commit and write test case for this bug.

### What issues does this PR fix or reference?
#56328

### Previous Behavior
below code work synchronously
```
__salt__['cassandra_cql.cql_query_with_prepare'](query,
                                                         'returner_return',
                                                         tuple(statement_arguments),
                                                         asynchronous=True)
```

### New Behavior
above code work asynchronously

### Tests written?
Yes

### Commits signed with GPG?

Yes

GwiYeong committed:
add test case for cassandra_cql async param

GwiYeong committed:
Revert "Support original API"

This reverts commit 377fbbc7b760bad1fd267747bf463e7cc8f3c1ae.

Ch3LL commented on salt/modules/cassandra_cql.py:
```diff
@@ -407,8 +411,13 @@ def cql_query_with_prepare(query, statement_name, statement_arguments, callback_
+    # Backward-compatibility. This should be removed in Sodium version
+    if 'async' in kwargs:
+        asynchronous = bool(kwargs['async'])
+        msg = 'Use of "async" parameter detected. The "asynchronous" parameter should be used instead.'
```
I dont think we should deprecate async, since this is such a small change. So we can ensure users dont need to update their states. Can we remove the deprecation messages?

GwiYeong replied:
ok, this message removed

Ch3LL commented on tests/unit/modules/test_cassandra_cql.py:
```diff
@@ -79,3 +79,19 @@ def test_unspecified_opts(self):
+        }
+
+        with patch.dict(cassandra_cql.__context__, mock_context):
+            cassandra_cql.cql_query_with_prepare('SELECT now() from system.local;', 'select_now', [], asynchronous=True)
```
can we also ensure `async` works as well here?

GwiYeong replied:
I add test for ensuring `async` works. 
but since python 3.7, async is reserved keyword so I have no choice but to pass async param as dict.

Ch3LL commented:
also looks like you have a legit test failure as well

GwiYeong committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

GwiYeong committed:
remove warning message
add test for async args

GwiYeong commented:
I don't know why the test is failed.
this is the failure log.
```
Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/unit/modules/test_cassandra_cql.py", line 96, in test_valid_async_args
    cassandra_cql.cql_query_with_prepare('SELECT now() from system.local;', 'select_now', [], asynchronous=True)
  File "/tmp/kitchen/testing/salt/modules/cassandra_cql.py", line 441, in cql_query_with_prepare
    session.row_factory = dict_factory
NameError: global name 'dict_factory' is not defined
```
`dict_factory` is imported from `cassandra.query`.  this is prerequisite for import `cassandra_cql` module, so if import error occurred for dict_factory, this test should not be run. 
how can I fix this import error..?
In my local environments, this test is working fine.

GwiYeong committed:
mocking cassandra.query.dict_factory to avoid test failure

GwiYeong commented:
@Ch3LL 
I tried to mock `from cassandra.query import dict_factory` but.. i don't know how to mock this statement. could you upgrade `cassandra-driver` version in the test environment?
`dict_factory` is moved from cassandra.decoder to cassandra.query. maybe `cassandra-driver` in the test environment is 1.x.
this is document about `dict_factory`. [doc](https://docs.datastax.com/en/developer/python-driver/3.21/api/cassandra/query/)
```
cassandra.query.dict_factory (colnames, rows)
Returns each row as a dict.

Example:

>>> from cassandra.query import dict_factory
>>> session = cluster.connect('mykeyspace')
>>> session.row_factory = dict_factory
>>> rows = session.execute("SELECT name, age FROM users LIMIT 1")
>>> print rows[0]
{u'age': 42, u'name': u'Bob'}
Changed in version 2.0.0: moved from cassandra.decoder to cassandra.query
```

Ch3LL commented:
looks like your right. I pulled down your pr and was able to run the tests just fine. Weirdly I'm having a hard time finding out how we are installing that requirement. I dont see cassandra in our requirements directory or in our https://github.com/saltstack/salt-jenkins repo which is where we install system dependencies. 

ping @s0undt3ch any ideas on how cassandra is even getting installed? I'm seeing the tests pass on jenkins: https://jenkinsci.saltstack.com/job/pr-centos7-py3/job/master/813/testReport/unit.modules.test_cassandra_cql/CassandraCQLReturnerTestCase/  so they aren't getting skipped and the test does try to import cassandra so this package is getting installed somehow. 

I think we should add this to our requirements directly instead, but still want to understand first how this is working currently.

s0undt3ch commented:
I see nowhere `cassandra-driver` being installed. Might we be mocking the module?

s0undt3ch commented:
@gwiyeong, how are you running tests?

GwiYeong commented:
@s0undt3ch 
I didn't do anything now... I don't know how cassandra is installed in test environments...
should I mock all of modules in cassandra?

s0undt3ch commented:
The jenkins builds are gone now.
There's also conflicts that need to be solved.
Could you fix those and push so we can get another set of builds please?

GwiYeong committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

GwiYeong commented:
@s0undt3ch 
I just fixed conflict from master. these conflicts are just code formatting.

GwiYeong committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

dwoz committed:
Blacken changed files

dwoz committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

Ch3LL commented:
@s0undt3ch i dont believe we are mocking the module. When i tested it it skips the test since the module is not installed. Even when i build the centos-py3 vm that jenkins uses as well. I also thought maybe the package was getting installed in an earlier test but i do not see that happening as well.

Lets try this @GwiYeong -> Can you add the cassandra requirement in this file: `requirements/static/linux.in` and when you run pre-commit it will run the tool that will compile all the requirements files for the tests.

s0undt3ch commented:
My guess is that @GwiYeong has cassandra already installed?
Ah, wait, `AttributeError: <module 'cassandra' from '/tmp/kitchen/testing/salt/modules/cassandra.py'> does not have the attribute 'query'`, rename `salt/modules/cassandra.py` to ``salt/modules/cassandra_mod.py`, on `__virtualname__` set it to `cassandra`, cleanup any `__pycache__` directories and `.pyc` files and run tests again.
That will fix it.

GwiYeong committed:
add cassandra-driver at requirements/static/linux.in
change virtualname for salt/module/cassandra

GwiYeong commented:
@s0undt3ch oh you are right. I also think `from cassandra import query` is try to import from`salt/modules/cassandra` instends of cassandra module from `cassandra-driver`.  so `salt/modules/cassandra` doesn't have `query module`, test is failed... 
`salt/modules/cassandra` is use pycassa that is maybe deprecated because [documents](https://github.com/pycassa/pycassa) says `it is highly recommended that you use the newer DataStax python driver instead of pycassa`

hmm. i changed **virtualname** of `salt/modules/cassandra` to `cassandra_mod`. is it ok..? no one use `salt/modules/cassandra`?? 

there is already tests for `salt/modules/cassandra`.. if this module doesn't use any more, you can remove this module and test cases from salt.

@Ch3LL  i added cassandra-driver in `salt/requirements/linux.in` for running cassandra test.

Ch3LL commented:
@s0undt3ch nice find.

@GwiYeong do you have pre-commit setup? When you edit the requirements/static files it should generate the appropriate appropriate requirements/static/py3{5,6,7} files for the tests. If you dont want to run pre-commit the diff is included in the jenkins output here: https://jenkinsci.saltstack.com/job/pr-pre-commit/job/PR-56404/7/console

GwiYeong committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

GwiYeong committed:
add missing asynchronous args

GwiYeong commented:
@Ch3LL 
I'm sorry. I missed that you said `when you run pre-commit it will run the tool that will compile all the requirements files for the tests.` 
but I don't know what is pre-commit setup means?  how to run pre-commit?
I would appreciate it if you let me know
thanks

Ch3LL commented:
Heres the documentation on how to setup pre-commit: https://docs.saltstack.com/en/latest/topics/development/contributing.html#quickstart  pre-commit will run a couple of different tools against your code when you run `git commit`. One of those tools checks to see if any requirement files have been changed and generates the needed files to add those dependencies for the tests.

If you dont want to set up pre-commit you can copy/paste the diff from the jenkins output and apply that instead.

GwiYeong committed:
add diff from jenkins instend running pre-commit

GwiYeong committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

GwiYeong committed:
add diff from jenkins instend of running pre-commit after merged with
master

GwiYeong commented:
@Ch3LL 
I just added the diff from jenkins output instead of running pre-commit.

thanks for your help

Ch3LL commented on salt/modules/cassandra.py:
```diff
@@ -43,7 +43,7 @@ def __virtual__():
 
     if HAS_PYCASSA and salt.utils.path.which("nodetool"):
-        return "cassandra"
+        return "cassandra_mod"
```
from @s0undt3ch 's comment:  "rename salt/modules/cassandra.py to salt/modules/cassandra_mod.py, on `__virtualname__` set it to cassandra"

So we will need to rename the file and set `__virtualname__` to cassandra.

Also it looks like there are some mac/windows failures.

GwiYeong replied:
i fixed this.

GwiYeong committed:
changed name salt/module/cassandra.py to salt/module/cassandra_mod.py and set its __virtualname__ is 'cassandra'
add cassandra to mac, window requirements.

GwiYeong committed:
fix for pre-commit
change cassandra test module name

xeacott committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

xeacott commented:
Im gunna give these a fresh run. Everything else looks good here.

xeacott commented:
@GwiYeong Anything else this needs? Otherwise i'll mark as merge ready.

GwiYeong commented:
@xeacott nothing to do anymore. thanks.

xeacott committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

xeacott committed:
Merge branch 'master' into fix-mismatch-function-args-cassandra-cql-return

xeacott committed:
Fix req files

xeacott committed:
Add reqs again
